Features/phase2

rudrbansal opened the pull request.

rudrbansal committed:
- Basic setup done
- Utilities and helper classes added
- UI of home view and detail view
- Business layer setup

rudrbansal committed:
Presenter file added
- Business logic moved to presenter file

rudrbansal committed:
Verbose code removed

rudrbansal committed:
Gemfile added
Pods deintegrated

rudrbansal committed:
- Added test files
- Code optimisation

rudrbansal committed:
- Pagination integrated
- Icon for pokemon added
- Code optimisation
- Extra files removed

rudrbansal committed:
- Home view cell modified
- Code optimisation

rudrbansal committed:
Delete Gemfile.lock

rudrbansal committed:
Delete Breakpoints_v2.xcbkptlist

trident10 commented on PokeDemo/Controllers/Base/BaseViewController.swift:
```diff
@@ -0,0 +1,18 @@
+import UIKit
+
+class BaseViewController: UIViewController {
+    
```
why this class is required?

rudrbansal replied:
This is made to write common code for all controllers.

trident10 replied:
Why not Protocol then?

rudrbansal replied:
Done

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    // MARK: - Private
+
+    private let presenter = HomeViewPresenter(service: Service.shared)
```
default shared parameter can be passed.

```
init(service: Service = Service.shared)
```
and thus can be used simply as - 
```
HomeViewPresenter()
```

rudrbansal replied:
Done in commit cd97b93468f78a9f35b47f3476ee786dce2aca8a

trident10 replied:
Next time create commit only with changes mentioned in the comment. Your mentioned commit doing so many unrelated changes which is not part of this comment

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+
+import UIKit
+
+class HomeViewController: BaseViewController {
```
Can be renamed as PokemonListViewController

rudrbansal replied:
Done in ba4937222566cb0f4e5ede9deb2e3a780b70e5d2

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    // MARK: - IBOutlets
+    
+    @IBOutlet weak var pokemonTableView: UITableView!
```
Can be renamed as tableView

rudrbansal replied:
Done in ba4937222566cb0f4e5ede9deb2e3a780b70e5d2

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    // MARK: - Private
+
+    private let presenter = HomeViewPresenter(service: Service.shared)
+    private var pokemonList: [Pokemon]? = [Pokemon]()
```
Can be renamed as `pokemons`

rudrbansal replied:
Done in 6f1d95bb7ca89332952ecd23599b4c8834939145

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        pokemonTableView.register(UINib(nibName: PokemonTableViewCell.cellIdentifier(), bundle: nil), forCellReuseIdentifier: PokemonTableViewCell.cellIdentifier())
```
Can be renamed as `PokemonListCell`

rudrbansal replied:
Done in ec81c95097ca23d12d71ba906f27718255168054

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+        super.viewDidLoad()
+        pokemonTableView.register(UINib(nibName: PokemonTableViewCell.cellIdentifier(), bundle: nil), forCellReuseIdentifier: PokemonTableViewCell.cellIdentifier())
+        presenter.setViewDelegate(delegate: self)
+        presenter.getPokemonList()
```
Since architecture is MVP. view **notifies** to presenter instead of giving **command**. Here `getPokemonList` behaves as view is giving a command to presenter to do something. Rather view must use notification sort of function names. Such as `viewLoaded` or `viewDidLoad`

```
presenter.viewDidLoad()
```

rudrbansal replied:
Done in 6f1d95bb7ca89332952ecd23599b4c8834939145

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+
+// MARK: - Extensions
+
+extension HomeViewController: UITableViewDataSource, UITableViewDelegate {
```
Divide the extension with separate protocol conformance. 

```
extension HomeViewController: UITableViewDataSource
```

```
HomeViewController: UITableViewDelegate
```

rudrbansal replied:
Done in b692a34b7b2c496b64ffcb60893e180e0ad736ea

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: PokemonTableViewCell.cellIdentifier()) as! PokemonTableViewCell
```
Handle force unwrap, if `dequeReusableCell` returns nil, it will crash while type casting to `PokemonTableViewCell`

rudrbansal replied:
As discussed, this will crash only if there is some development issue. It will work in case there is no data to display.

trident10 replied:
Make a habit not to crash at all. Or you really want to then use assert that crash only for debug builds and go silent in release

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: PokemonTableViewCell.cellIdentifier()) as! PokemonTableViewCell
+        let pokemon = pokemonList![indexPath.row]
```
Handle force unwrap, if `pokemonList` returns nil, it will crash

rudrbansal replied:
Done in b692a34b7b2c496b64ffcb60893e180e0ad736ea

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+        let cell = tableView.dequeueReusableCell(withIdentifier: PokemonTableViewCell.cellIdentifier()) as! PokemonTableViewCell
+        let pokemon = pokemonList![indexPath.row]
+        cell.setupData(data: pokemon)
+        presenter.getPokemonInfo(pokemon: pokemon) { image in
```
Since architecture is MVP. view notifies to presenter instead of giving command. Here `getPokemonInfo` behaves as view is giving a command to presenter to do something. Rather view must use notification sort of function names. Such as `didSetupCell(with: pokemon)`

Consider view should not know what presenter is going to do with this information. View's job is to provide status to presenter that something happened at view and presenter will use these methods, to download image or do something which is not related to view. Thus it separates the concerns to specific layer of the architecture. Also, does not bleed out the information to different layers

trident10 replied:
So if you use the simple method didSetupCell, you might need to pass some unique cell info, such as row or something, so that when presenter downloads the image it can call a separate function to view saying, `view.displayImage(image, at: indexRow)`

rudrbansal replied:
Done in 6f1d95bb7ca89332952ecd23599b4c8834939145

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    }
+    
+    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        let lastRowIndex = tableView.numberOfRows(inSection: 0) - 1
+        if indexPath.row == lastRowIndex {
+            let spinner = UIActivityIndicatorView(style: .medium)
+            spinner.startAnimating()
+            spinner.frame = CGRect(x: CGFloat(0), y: CGFloat(0), width: tableView.bounds.width, height: CGFloat(44))
+            if presenter.isLoadingCell(for: indexPath) {
+                self.pokemonTableView.tableFooterView = spinner
+                self.pokemonTableView.tableFooterView?.isHidden = false
+                presenter.getPokemonList()
+            } else {
+                self.pokemonTableView.tableFooterView?.isHidden = true
+            }
+        }
```
view is handling the logic. The logic should be covered in the presenter. Remove it and do it in separate PR

rudrbansal replied:
Done in cd97b93468f78a9f35b47f3476ee786dce2aca8a

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+
+extension HomeViewController: HomeViewPresenterDelegate {
+    
+    func showPokemonList(pokemons: [Pokemon]) {
```
Rename to `func show(_ pokemons: [Pokemon])`

rudrbansal replied:
Done in 6f1d95bb7ca89332952ecd23599b4c8834939145

trident10 commented on PokeDemo/Controllers/HomeViewController.swift:
```diff
@@ -0,0 +1,91 @@
+    
+    func showPokemonList(pokemons: [Pokemon]) {
+        pokemonList?.append(contentsOf: pokemons)
+        pokemonTableView.dataSource = self
+        pokemonTableView.delegate = self
```
It should be set only `ONCE`. Must set while initiating this viewController or once `tableView` is initiated

rudrbansal replied:
Done in b692a34b7b2c496b64ffcb60893e180e0ad736ea

trident10 commented:
Use commit style guide -  https://gist.github.com/ericavonb/3c79e5035567c8ef3267

Commit atomic code change (means the minimum code change that fix the smallest change).
Commit often not deload everything in one commit.

trident10 commented on PokeDemo/CustomViews/TableViewCells/PokemonTableViewCell.swift:
```diff
@@ -0,0 +1,26 @@
+    
+    // MARK: - Outlets
+    
+    @IBOutlet weak var lblTitle: UILabel!
```
can be renamed as title

rudrbansal replied:
Done in 318a882a3ad1532005f3e532a0215cdf07ec67ed

trident10 commented on PokeDemo/CustomViews/TableViewCells/PokemonTableViewCell.swift:
```diff
@@ -0,0 +1,26 @@
+    // MARK: - Outlets
+    
+    @IBOutlet weak var lblTitle: UILabel!
+    @IBOutlet weak var iconImageView: UIImageView!
```
can be renamed as icon

rudrbansal replied:
Done in 318a882a3ad1532005f3e532a0215cdf07ec67ed

trident10 commented on PokeDemo/CustomViews/ViewBaseClasses/BaseTableViewCell.swift:
```diff
@@ -0,0 +1,17 @@
+
+import UIKit
+
+class BaseTableViewCell: UITableViewCell {
```
Why this is required?

rudrbansal replied:
This is made to write common code for all table view cells.

trident10 replied:
It is recommended to use protocol instead of inheritance. Refer this - https://medium.com/bleeding-edge/nicer-reuse-identifiers-with-protocols-in-swift-97d18de1b2df

rudrbansal replied:
Covered in 32fb355207022acc0d4f94754537ed669f817815

trident10 commented on PokeDemo/Extensions/UIView.swift:
```diff
@@ -0,0 +1,85 @@
+
+import UIKit
+
+extension UIView {
```
why it is required?

rudrbansal replied:
Added IBInspectable attributes here for views.

trident10 replied:
Why?

rudrbansal replied:
Covered in 2d053f9c05f3152552080a0011a49ee110a1efc8

trident10 commented on PokeDemo/Models/Pokemon.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+struct Pokemon: Decodable, Equatable {
```
why Equatable?

rudrbansal replied:
It is done to check for data in unit test cases.

trident10 replied:
Then move it in unit test only.

trident10 replied:
Move it in unit test target

rudrbansal replied:
Covered in dc6c3b12b54f65c2219d3117fb365e29511a26ee

trident10 commented on PokeDemo/Models/Pokemon.swift:
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+struct Pokemon: Decodable, Equatable {
+    var name: String?
+    var url: String?
```
why optionals?

rudrbansal replied:
These are coming in response, so kept them optional.

trident10 replied:
Remove optionals. If any response does not come with value handle as JSON decode error

rudrbansal replied:
Covered in 8e62e9f9cf6903a304d206bb02281454063764a0

trident10 commented on PokeDemo/Models/Sprite.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+struct Sprite: Decodable {
```
What is Sprite?

rudrbansal replied:
All pokemon detail is returned in response with this key, so kept it's name same

trident10 replied:
Change the name. It doesnot mean to our project models. Learn about domain models. Check article in JSON decoder to keep separate names that doesn't match JSON response key

rudrbansal replied:
Covered in f05532612b67d77b533e2e3503714f79ff96df2d

trident10 commented on PokeDemo/Presenter/HomeViewPresenter.swift:
```diff
@@ -0,0 +1,105 @@
+    // MARK: Public
+    
+    weak var delegate: HomePresenterDelegate?
+    private let errorTitle = "Error"
```
Move under Private MARK

rudrbansal replied:
Done

trident10 replied:
Where?

rudrbansal replied:
Covered in cd97b93468f78a9f35b47f3476ee786dce2aca8a

trident10 commented on PokeDemo/Presenter/HomeViewPresenter.swift:
```diff
@@ -0,0 +1,105 @@
+        self.delegate = delegate
+    }
+    
+    func isLoadingCell(for indexPath: IndexPath) -> Bool {
```
what is it used for?

rudrbansal replied:
It was used to check if api should get more results. Removed for now.

trident10 replied:
> It was used to check if api should get more results. Removed for now.

Where did you do it?

rudrbansal replied:
cd97b93468f78a9f35b47f3476ee786dce2aca8a

trident10 commented on PokeDemo/Presenter/HomeViewPresenter.swift:
```diff
@@ -0,0 +1,105 @@
+    }
+    
+    func getPokemonList() {
+        if let nextURL {
+            self.nextURL = nextURL
+        } else {
+            nextURL = Endpoints.shared.baseURL
+        }
```
Remove this, lets do pagination in next PR

rudrbansal replied:
Done

trident10 replied:
> Done

I don't see any commit number of this change

rudrbansal replied:
Done in cd97b93468f78a9f35b47f3476ee786dce2aca8a

trident10 replied:
[cd97b93](https://github.com/rudrbansal/Pokemon/pull/2/commits/cd97b93468f78a9f35b47f3476ee786dce2aca8a)

rudrbansal committed:
refactor: remove extra code
remove pagination logic
add constants file

rudrbansal committed:
refactor: rename IBoutlets

Rename of IBOutlets in PokemonTableViewCell

closes PR comment

rudrbansal committed:
refactor: remove CommonUtilities file

remove CommonUtilities file and class and move code to base view controller

closes PR comment

rudrbansal committed:
refactor: change class name and controller name

change name of HomeViewController

closes PR comment

rudrbansal committed:
refactor: rename class

change cell class name and file name change name of PokemonTableViewCell

closes PR comment

rudrbansal committed:
refactor: rename presenter

rename presenter and functions in presenter

closes PR comment

rudrbansal committed:
refactor: error handling

remove forced unwrapping of pokemons array
segregation of data source and delegate
table view data source and delegates set on view did load

closes PR comments

trident10 commented on PokeDemo/Controllers/PokemonListViewController.swift:
```diff
@@ -0,0 +1,80 @@
+    
+    // MARK: - IBOutlets
+    
+    @IBOutlet weak var tableView: UITableView!
```
**question:** 
1. why weak?
2. Can it be private?

rudrbansal replied:
Covered in feda558a80950ca51ca6477354446339522b97cd
It is weak as outlets should lose memory when view is gone

trident10 commented on PokeDemo/Controllers/PokemonListViewController.swift:
```diff
@@ -0,0 +1,80 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        tableView.register(UINib(nibName: PokemonListCell.cellIdentifier(), bundle: nil), forCellReuseIdentifier: PokemonListCell.cellIdentifier())
+        tableView.dataSource = self
+        tableView.delegate = self
```
**suggestion:** Instead of dumping all sorts of configurations in `viewDidLoad`, you can leverage the property observers -  [Refer here](https://docs.swift.org/swift-book/LanguageGuide/Properties.html)

```
@IBOutlet weak var tableView: UITableView! {
    didSet {
        tableView.register(UINib(nibName: PokemonListCell.cellIdentifier(), bundle: nil), forCellReuseIdentifier: PokemonListCell.cellIdentifier())
        tableView.dataSource = self
        tableView.delegate = self
 } 
}
```

rudrbansal replied:
Covered in 65ba611f7b38b56cf726f9afed6a8efc55e5e8b2

trident10 commented on PokeDemo/Controllers/PokemonListViewController.swift:
```diff
@@ -0,0 +1,80 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: PokemonListCell.cellIdentifier()) as! PokemonListCell
```
**issue:** Force unwrap can crash if cellIdentifier is changed. Someone changed cellIdentifier in storyboard.

rudrbansal replied:
Covered in 80b1ad6d15724ee26a476310a11c9bffe00a6cd9

trident10 commented on PokeDemo/Controllers/PokemonListViewController.swift:
```diff
@@ -0,0 +1,80 @@
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: PokemonListCell.cellIdentifier()) as! PokemonListCell
+        if let pokemons {
```
use guard instead, It will avoid nested blocks and give more readability

rudrbansal replied:
Covered in 9ed73dd684475e197199a93afe12e7284683f8ac

trident10 commented:
https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

trident10 commented on PokeDemo/Controllers/PokemonListViewController.swift:
```diff
@@ -0,0 +1,80 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: PokemonListCell.cellIdentifier()) as! PokemonListCell
+        if let pokemons {
+            let pokemon = pokemons[indexPath.row]
```
what happens if pokemons array is updated and indexpath.row uses the old value? It may cause out of bound array crash

rudrbansal replied:
Covered in 9f8ab188f324bcce4e3d736aa3709cafd8996268

trident10 commented on PokeDemo/Controllers/PokemonListViewController.swift:
```diff
@@ -0,0 +1,80 @@
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let detailViewController: PokemonDetailViewController = moveToViewController(storyboard: "Main", destination: PokemonDetailViewController.identifier()) as! PokemonDetailViewController
+        detailViewController.pokemon = (pokemons?[indexPath.row])
```
Remove Pokemon detail  related code entirely from this PR

rudrbansal replied:
Covered in 354546edd83968c52af819d1834b0a0f8e6d7f93

trident10 commented on PokeDemo/CustomViews/TableViewCells/PokemonListCell.swift:
```diff
@@ -0,0 +1,26 @@
+    // MARK: - Exposed Methods
+
+    func setupData(data: Pokemon) {
+        title.text = data.name?.uppercased()
```
not sure what happens when name is nil. What will it show as title?

rudrbansal replied:
It will show blank in case name is nil

rudrbansal committed:
REFACTOR: Error handling

Use guard let instead of force unwrapping.

Closes PR comment.

rudrbansal committed:
REFACTOR: Access specifiers

Set outlets as private

Closes PR comment.

rudrbansal committed:
REFACTOR: Remove extra code

Remove delegate method which is part of future scope

Closes PR comment

rudrbansal committed:
REFACTOR: Remove base class

Move reusable code from base table view cell to protocol and implement that protocol to table view cell extension

Closes PR comment

rudrbansal committed:
REFACTOR: Error handling

Use guard let instead of if let.

Closes PR comment.

rudrbansal committed:
REFACTOR: Better placement of code

Set table view data source and register cell in didset instead of view did load. Removes extensive burden from view did load

Closes PR comment

rudrbansal committed:
REFACTOR: Error handling

Check if index path is less than number of rows. This check is added to avoid crash

Closes PR comment

rudrbansal committed:
REFACTOR: Remove file

Remove IBInspectable properties to improve performance

Closes PR comment

rudrbansal committed:
REFACTOR: Remove code

Use protocol to get identifier instead of using function in base

BREAKING CHANGE: No need to write multiple functions to achieve same result

rudrbansal committed:
REFACTOR: Set class as final

Set classes which should not be inherited as final

BREAKING CHANGE: Improves performance

rudrbansal committed:
REFACTOR: Remove future scope

Remove future scope requirements from current branch

Closes PR comment

rudrbansal committed:
REFACTOR: Remove future scope

Remove future scope presenter from current branch

Closes PR comment

rudrbansal committed:
REFACTOR: Remove nesting of conditions

Use guard let instead of if let to avoid nesting of conditions

Closes PR comment

trident10 commented on PokeDemo/Controllers/PokemonListViewController.swift:
```diff
@@ -0,0 +1,73 @@
+    // MARK: - Private
+    
+    private let presenter = PokemonListPresenter()
+    private var pokemons: [Pokemon]? = [Pokemon]()
```
```suggestion
    private var pokemons: [Pokemon] = []
```

rudrbansal replied:
Covered in c3f06092d44508dfe9c0782dff1248c27cb2cd05

trident10 commented on PokeDemo/Controllers/PokemonListViewController.swift:
```diff
@@ -0,0 +1,73 @@
+extension PokemonListViewController: UITableViewDataSource {
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return pokemons?.count ?? 0
```
pokemons.count

rudrbansal replied:
Covered in c3f06092d44508dfe9c0782dff1248c27cb2cd05

trident10 commented on PokeDemo/Controllers/PokemonListViewController.swift:
```diff
@@ -0,0 +1,73 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: PokemonListCell.reuseIdentifier) as? PokemonListCell else { return UITableViewCell() }
+        guard let pokemons else { return PokemonListCell() }
+        if indexPath.row <= pokemons.count {
+            cell.setupData(data: pokemons[indexPath.row])
+            presenter.didSetupCellWith(pokemon: pokemons[indexPath.row]) { image in
+                if let image = image{
+                    DispatchQueue.main.async {
+                        cell.setupImage(image: image)
+                    }
+                }
+            }
+        }
+        return cell
+    }
```
could be much cleaner

```
guard let cell = tableView.dequeueReusableCell(withIdentifier: PokemonListCell.reuseIdentifier) as? PokemonListCell,
guard indexPath.row <= pokemons.count else { UITableViewCell() }
cell.setupData(data: pokemons[indexPath.row])
presenter.didSetupCellWith(pokemon: pokemons[indexPath.row]) { image in
    if let image = image {
        DispatchQueue.main.async {
            cell.setupImage(image: image)
        }
    }
}
return cell
```

rudrbansal replied:
Covered in 7416c016974e6b9ad628b5d993d1f038512fa7a9

trident10 commented on PokeDemo/Helpers/Internet.swift:
```diff
@@ -0,0 +1,52 @@
+import Foundation
+import SystemConfiguration
+
+public protocol InternetManager {
```
We don't need this. Remove InternetManager. If we need this specific use case, we'll introduce it later. As of now if there is no internet, url connection will send time out error response.

rudrbansal replied:
Covered in 7db07cefc3d5401a459807214b87c0e8202a8f7e

trident10 commented on PokeDemo/Models/PokemonResult.swift:
```diff
@@ -0,0 +1,15 @@
+import Foundation
+
+struct PokemonResult: Decodable {
+    var count: Int?
```
Remove optionals

rudrbansal replied:
Covered in 4bf40a3f8029568b019914197f8773b6da2e3d75. Previous is kept optional as it is coming null

trident10 commented on PokeDemo/Models/Sprite.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+struct Sprite: Decodable {
+    var name: String?
+    var id: Int?
+    var sprites: SpriteDetail?
+}
```
Remove optionals

rudrbansal replied:
Covered in 37f42463fea0c21c150bf660477077e13b91ba3a

trident10 commented on PokeDemo/Presenter/PokemonListPresenter.swift:
```diff
@@ -0,0 +1,97 @@
+    func showAlert(title: String, message: String)
+}
+
+typealias PokemonListViewPresenterDelegate = PokemonListPresenterDelegate & UIViewController
```
It does not make any sense to introduce at this stage. Remove

rudrbansal replied:
Covered in e5b038e391ff3c33386eaab59afd64089109162d

trident10 commented on PokeDemo/Presenter/PokemonListPresenter.swift:
```diff
@@ -0,0 +1,97 @@
+        getPokemons()
+    }
+    
+    func setViewDelegate(delegate: PokemonListViewPresenterDelegate) {
+        self.delegate = delegate
+    }
```
delegate property is already exposed. Either make delegate property private or remove this function

rudrbansal replied:
Covered in 4375f7220db9a1b8d049f766932eeccb11364e38

trident10 commented on PokeDemo/Presenter/PokemonListPresenter.swift:
```diff
@@ -0,0 +1,97 @@
+    }
+    
+    func didSetupCellWith(pokemon: Pokemon, completion: @escaping (UIImage?) -> Void) {
+        service.sendRequestWithJSON(endpoint: pokemon.url!, method: .get) { response, error in
```
if url is nil, it will crash

rudrbansal replied:
Covered in 8e62e9f9cf6903a304d206bb02281454063764a0

trident10 commented on PokeDemo/Presenter/PokemonListPresenter.swift:
```diff
@@ -0,0 +1,97 @@
+    
+    func didSetupCellWith(pokemon: Pokemon, completion: @escaping (UIImage?) -> Void) {
+        service.sendRequestWithJSON(endpoint: pokemon.url!, method: .get) { response, error in
+            if error == nil {
```
You can use guard error == nil to avoid the nested blocks. Too messy to read nested  blocks. Avoid not to nest blocks as much as possible

rudrbansal replied:
Covered in c22077d2780463280a58fb8b17420e0080b6decf

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+
+import XCTest
+@testable import PokeDemo
+import Alamofire
```
Why Alamofire is needed?

rudrbansal replied:
Covered in 6ab5f663a5d4c54348ba7c497892c3722d9ca9a0

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+    
+    override func setUp() {
+        super.setUp()
+        presenter = PokemonListPresenter(service: Service.shared)
```
You should not be injecting real Service object. Rather inject MockService object

rudrbansal replied:
Covered in 95b810a9d180840d001ce309cd07301a9c2284de

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+    
+    func testGetPokemonListCallsAPI(){
+        // Given
+        let mockService = MockService()
+        
+        // When
+        presenter = PokemonListPresenter(service: mockService)
```
If you inject MockService in setup. You don't need to do it in every tests. 

Read XCTest lifecycle - 

http://trikalabs.com/xctestcase-lifecycle/

trident10 replied:
Or better as mentioned in the block, you can setup presenter in class setup and deinit in class teardown

rudrbansal replied:
Covered in 95b810a9d180840d001ce309cd07301a9c2284de

trident10 commented on PokeDemoTests/MockService.swift:
```diff
@@ -0,0 +1,40 @@
+
+final class MockService: ServiceRepresentable {
+    
+    var sendRequestWithJSONIsCalledIndex = 0
+    var sendRequestWithJSONEndPoint: String = ""
```
Create these properties read only. We don't want that in tests these values can be overwritten by someone. Its value must be changed only by the methods which are called internally for which we are writing tests

trident10 replied:
suggestion - var sendRequestWithJSONCallCount = 0

trident10 replied:
You need to create read only variables for method, parameter and header too. To test right values are passed for these as well or not.

rudrbansal replied:
Covered in 9bbf1c08498fe5f761b27131dee55b2fd25072cd

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+        // Then
+        XCTAssertEqual(mockService.sendRequestWithJSONIsCalledIndex, 1)
+        XCTAssertEqual(mockService.sendRequestWithJSONEndPoint, "https://pokeapi.co/api/v2/pokemon")
+    }
```
add asserts that right method, parameters, header are passed and completion is not nil

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+        XCTAssertEqual(mockService.sendRequestWithJSONEndPoint, "https://pokeapi.co/api/v2/pokemon")
+    }
+    
+    func testGetPokemonListReturnsSuccessFromAPI(){
```
Remove FromAPI

rudrbansal replied:
Covered in 9d1c5ccff0f3599a5ee73c4d5954637102b2f810

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+//        mockService.onCompletion?(stubPokemons, nil)
+        
+        // Then
+        XCTAssertEqual(mockDelegate.pokemons, stubPokemons)
```
Assert - mockDelegate show function call count as well

rudrbansal replied:
Done

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+        mockService.onCompletion?(nil, TestError())
+        
+        // Then
+        XCTAssertEqual(mockDelegate.showAlertCalled, true)
```
Test against call count

rudrbansal replied:
Covered in d703a43a6c677facf19eed61471ae0cdd9a3482f

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+        XCTAssertEqual(mockDelegate.showAlertCalled, true)
+        XCTAssertEqual(mockDelegate.showAlertTitle, Constants.error)
+        XCTAssertEqual(mockDelegate.showAlertMessage, "The operation couldn’t be completed. (PokeDemoTests.TestError error 1.)")
+        // Check the mockDelegate showPokemonList function is called with right Pokemon array
```
Remove comments

rudrbansal replied:
Done

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+        // Check the mockDelegate showPokemonList function is called with right Pokemon array
+    }
+    
+    func testGetPokemonListReturnsNoInternetFailureFromAPI(){
+        // Given
+        let mockService = MockService()
+        let mockDelegate = MockPresenterDelegate()
+        presenter = PokemonListPresenter(service: mockService)
+        presenter.setViewDelegate(delegate: mockDelegate)
+        presenter.viewDidLoad()
+        
+        // When service sends successful response
+        mockService.onCompletion?(nil, TestError())
+        
+        // Then
+        XCTAssertEqual(mockDelegate.showAlertCalled, true)
+        XCTAssertEqual(mockDelegate.showAlertTitle, Constants.error)
+        XCTAssertEqual(mockDelegate.showAlertMessage, Internet.shared.noInternet)
+    }
```
Isn't it the same test as previous. What's the difference?

rudrbansal replied:
Covered in 14abf11f45283b232c0d56311c6b93990f9ba842

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+    
+    func testGetPokemonListReturnsInvalidDataFromAPI(){}
+    
+    func testGetPokemonListReturnsGeneralErrorFromAPI(){
+        // Given
+        let mockService = MockService()
+        
+        // When
+        presenter = PokemonListPresenter(service: mockService)
+        presenter.viewDidLoad()
+        
+        // Then
+        XCTAssertEqual(mockService.sendRequestWithJSONIsCalledIndex, 1)
+    }
```
Remove this test

rudrbansal replied:
Covered in a7a16501021c48b8a1bbbf9c8a519a0146eeeb8c

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+        XCTAssertEqual(mockDelegate.showAlertMessage, Internet.shared.noInternet)
+    }
+    
+    func testGetPokemonListReturnsInvalidDataFromAPI(){}
```
Add this test implementation

rudrbansal replied:
Done

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+
+final class MockPresenterDelegate: PokemonListViewPresenterDelegate {
+        
+    var showAlertCalled: Bool = false
+    var showAlertTitle: String?
+    var showAlertMessage: String?
+    var pokemons: [Pokemon]?
```
Make it read only

rudrbansal replied:
Covered in 9bbf1c08498fe5f761b27131dee55b2fd25072cd

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+    var showAlertMessage: String?
+    var pokemons: [Pokemon]?
+    
+    func show(_ pokemons: [Pokemon]) {
+        self.pokemons = pokemons
+    }
```
Add call count variable

trident10 commented on PokeDemoTests/PokemonListPresenterTests.swift:
```diff
@@ -0,0 +1,142 @@
+    }
+    
+    func showAlert(title: String, message: String) {
+        showAlertCalled = true
```
Make it call count instead

rudrbansal replied:
Covered in d703a43a6c677facf19eed61471ae0cdd9a3482f

rudrbansal committed:
REFACTOR: Remove equatable

Remove equatable from Pokemon model

Closes PR comment

rudrbansal committed:
REFACTOR: Remove optionals

Remove optional from Pokemon model

Closes PR comment

rudrbansal committed:
REFACTOR: Change model names

Rename Sprite to PokemonAttributes and SpriteDetails to ImageAttributes

Closes PR comment

rudrbansal committed:
REFACTOR: Declare empty array

Declare empty array and remove optional from declaration

Closes PR comment

rudrbansal committed:
REFACTOR: Code optimisation

Code optimisation by use of guard statement

Closes PR comment

rudrbansal committed:
REFACTOR: Remove optionals

Remove optional from PokemonAttributes and ImageAttribute model

Closes PR comment

rudrbansal committed:
REFACTOR: Code optimisation

Code optimisation by use of guard statement

Closes PR comment

rudrbansal committed:
REFACTOR: Remove optionals

Remove optional from PokemonResult model

Closes PR comment

rudrbansal committed:
REFACTOR: Remove Internet class

Remove Internet class to check if Internet is available or not

Closes PR comment

rudrbansal committed:
REFACTOR: Code optimisation

Remove type alias of delegate

Closes PR comment

rudrbansal committed:
REFACTOR: Code optimisation

Set delegate as private

Closes PR comment

rudrbansal committed:
REFACTOR: Code optimisation

Remove setViewDelegate function

Closes PR comment

rudrbansal committed:
REFACTOR: Code optimisation

Remove alamofire from test cases

Closes PR comment

rudrbansal committed:
REFACTOR: Use common object of mockService

Use common object of mock service instead of initialising in every function

Closes PR comment

rudrbansal committed:
REFACTOR: Remove MockInternet

Remove MockInternet from current scope

Maintains clean code

rudrbansal committed:
REFACTOR: Code optimisation

Remove FromAPI from function names

Closes PR comment

rudrbansal committed:
REFACTOR: Set mockDelegate

Set mockDelegate directly instead of calling function

Maintains clean code

rudrbansal committed:
Bug fix: Make PokemonResult encodable

Make PokemonResult model as encodable in order to encode it

Fixes bug

rudrbansal committed:
Refactor: Remove test func for Internet

Remove unused test func for Internet

Closes PR comment

rudrbansal committed:
TEST: Check for empty array

Add test to check if response array is empty

Covers test case

rudrbansal committed:
REFACTOR: Remove unused func

Remove used test func for general error

Closes PR comment

rudrbansal committed:
REFACTOR: Move model to separate file

Move model PokemonImageAttribute to separate file from PokemonAttribute

Maintains clean code

rudrbansal committed:
BUG: JSONEncoder returning nil

JSONEncoder returns nil data when it encodes PokemonResult. Add coding keys to fix it.

Fixes bug

rudrbansal committed:
REFACTOR: Remove UIKit from presenter

Remove UIKit from presenter as presenter should be UIKit independent

Maintains single dependency

rudrbansal committed:
DOC: Update comments

Update comments for test functions

Maintains documentation

trident10 commented on PokeDemo/Constants.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+enum Constants {
```
```suggestion
enum Constant {
```

rudrbansal replied:
Done

trident10 commented on PokeDemo/Controllers/Base/BaseViewController.swift:
```diff
@@ -0,0 +1,20 @@
+    
+    // MARK: - Exposed Methods
+    
+    func moveToViewController(storyboard: String, destination: String) -> UIViewController {
+        let storyboard = UIStoryboard(name: storyboard, bundle: nil)
+        let vc = storyboard.instantiateViewController(withIdentifier: destination)
+        return vc
+    }
+}
```
Is it used anywhere?

rudrbansal replied:
Covered in cf0322761f261fbbe3e6c7997b28c2d417663ac9

trident10 commented on PokeDemo/CustomViews/TableViewCells/PokemonListCell.swift:
```diff
@@ -0,0 +1,26 @@
+    
+    // MARK: - Outlets
+    
+    @IBOutlet weak var title: UILabel!
```
weak not required

rudrbansal replied:
Covered in c141e033217fe7cb155df5baff1dbc9b44c0b899

trident10 commented on PokeDemo/Extensions/UITableViewCell.swift:
```diff
@@ -0,0 +1,21 @@
+//
+//  UITableViewCell.swift
```
Rename UITableViewCell+ReuseIdentifying

rudrbansal replied:
Covered in bf657192bcbf9c8546d097f1262e9261d6bac8d2

rudrbansal committed:
STYLE: Folder structure setup

Improvements in folder structure for better formatting

Updates folders

rudrbansal committed:
REFACTOR: Change response from any to data

Change response from any to data to avoid typecasting. Also, it avoids to receive any type of data

Improves performance

rudrbansal committed:
REARRANGE: Delete BaseViewController

Delete BaseViewController as it is out of current scope

Maintains clean code

rudrbansal committed:
REFACTOR: Remove do catch

Remove do catch statement from didSetupCell logic

Improves performance and clean code

rudrbansal committed:
REFACTOR: Move PokemonList view to new storyboard

Move PokemonList view to new storyboard and use storyboard reference in Main storyboard

Improves performance and maintains clean code

rudrbansal committed:
REFACTOR: Remove weak from outlets

Remove weak from outlets as it is not required anymore

Maintains clean code

rudrbansal committed:
REFACTOR: Change response type from Any to Data

Change response type from Any to Data to restrict it to get Data in response

Improves performance

rudrbansal committed:
REFACTOR: Remove loadJson func

Remove loaadJson func
Load data from local JSON

Improves code performance and reduces complexity

rudrbansal committed:
REFACTOR: Change test case logic

Use local JSON instead of getting JSON from file
Add index count for API count
Use common instance of presenter instead of making new instance

Improves performance

rudrbansal committed:
REFACTOR: Change test case logic

Add index count for API count

Maintains same logic for all cases

rudrbansal committed:
REFACTOR: Remove extra code

Remove given title when there is nothing in given
Use common instance of presenter instead of making new instance
Use "Error" text

Maintains clean code

trident10 commented:
Remove all ds_store files. Refer - https://stackoverflow.com/questions/107701/how-can-i-remove-ds-store-files-from-a-git-repository

trident10 commented on PokeDemo/Screens/PokemonList/Model/Pokemon.swift:
```diff
@@ -0,0 +1,19 @@
+    var name: String
+    var url: String
+    
+    enum CodingKeys: String, CodingKey {
+        case name
+        case url
+    }
```
Do we need CodingKeys?

rudrbansal replied:
Covered in a4731a787e76e1973e94509f7945ebccdc9fc4e1

trident10 commented on PokeDemo/Screens/PokemonList/Model/PokemonResult.swift:
```diff
@@ -0,0 +1,22 @@
+    var previous: String?
+    var results: [Pokemon]
+    
+    enum CodingKeys: String, CodingKey {
+        case count
+        case next
+        case previous
+        case results
+    }
```
Do we need this?

rudrbansal replied:
Covered in 04e0ac57f896573a4a6419d22c105cebd88f63b1

trident10 commented on PokeDemo/Screens/PokemonList/Model/Service.swift:
```diff
@@ -0,0 +1,49 @@
+    func sendRequestWithJSON(
+        endpoint: String,
+        method: HTTPMethod,
+        parameters: [String:Any]?,
```
Can it be [String: String] instead of [String: Any]?

rudrbansal replied:
Covered in 8c9007acbdcba689302ff066340d4f232a35f3f7

trident10 commented on PokeDemoTests/MockService.swift:
```diff
@@ -0,0 +1,29 @@
+
+final class MockService: ServiceRepresentable {
+    
+    var sendRequestWithJSONIsCalledIndex = 0
```
```suggestion
    var sendRequestWithJSONIsCalledCount = 0
```

rudrbansal replied:
Covered in 7d2c96535ac909ec36845ef65ed69923d4c66332

trident10 commented on PokeDemoTests/Resources/PokemonList.json:
```diff
@@ -0,0 +1,47 @@
+
```
Is it required now?

rudrbansal replied:
Covered in 37dab3f9f4885f47ac8e5f7d2809f2b5aeeb6fb9

rudrbansal committed:
REFACTOR: Change variable name

Change variable name from sendRequestWithJSONIsCalledIndex to sendRequestWithJSONIsCalledCount

Closes PR comment

rudrbansal committed:
REARRANGE: Remove JSON file

Remove JSON file from code

Closes PR comment

rudrbansal committed:
REFACTOR: Remove coding keys enum

Remove coding keys enum from Pokemon model

Closes PR comment

rudrbansal committed:
REFACTOR: Remove coding keys enum

Remove coding keys enum from Pokemon model

Closes PR comment

rudrbansal committed:
REARRANGE: Remove DS_Store

Remove DS_Store files from code

Closes PR comment

rudrbansal committed:
Refactor: Change parameter type

Change parameter type from [String:Any] to [String:String] to restrict from sending any type of data in parameters

Closes PR comment

rudrbansal commented:
> Remove all ds_store files. Refer - https://stackoverflow.com/questions/107701/how-can-i-remove-ds-store-files-from-a-git-repository

Covered in 5afb0bd7c4d72936f4565114763b268f5bb4614b

rudrbansal committed:
REFACTOR: Make read only properties

Make read only properties for Mock Service and Test class

Closes PR comment

rudrbansal committed:
REFACTOR: Change variable type

Change variable type from Bool to Int

Closes PR

rudrbansal committed:
REFACTOR: Add fetch pokemon count

Add fetch pokemon count when pokemons are fetched successfully to check that func is called only once

Closes PR comment

trident10 commented on PokeDemo/Screens/PokemonList/Model/Service.swift:
```diff
@@ -0,0 +1,49 @@
+    // MARK: - Singleton
+    
+    static let shared = Service()
+    
```
issue: Add private init to make it singleton

rudrbansal committed:
REFACTOR: Make presenter variable as optional

Make presenter variable as optional instead of force unwrapping

Improves code performance

rudrbansal committed:
TEST: Add more test cases

Add test case for nil response, success with error and wrong data

Increases code coverage

rudrbansal committed:
REFACTOR: Make service class as singleton

Make service class as singleton by setting it's init as private

Improves code performance

rudrbansal committed:
FEAT: Add Alamofire to test target

Add Alamofire to test target to access for test cases

Enables test case coverage for alamofire

rudrbansal committed:
TEST: Add asserts to check parameters

Add assets to check parameters such as methods, params and header

Increases code coverage
